crates/wit-encoder/src/function.rs

```diff
@@ -0,0 +1,177 @@
+        while let Some((name, type_)) = peekable.next() {
+            write!(f, "{}: {}", name, type_)?;
+            if peekable.peek().is_some() {
+                write!(f, ", ")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Params {
+    pub fn empty() -> Self {
+        Self::default()
+    }
+
+    pub fn items(&self) -> &Vec<(Ident, Type)> {
+        &self.items
+    }
+
+    pub fn items_mut(&mut self) -> &mut Vec<(Ident, Type)> {
+        &mut self.items
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+#[cfg_attr(feature = "serde", serde(untagged))]
+pub enum Results {
+    #[cfg_attr(feature = "serde", serde(serialize_with = "serialize_params"))]
+    Named(Params),
+    #[cfg_attr(feature = "serde", serde(serialize_with = "serialize_anon_result"))]
+    Anon(Type),
+}
+
+impl Display for Results {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Results::Anon(type_) => type_.fmt(f)?,
+            Results::Named(vals) => {
+                if !vals.items.is_empty() {
+                    write!(f, "(")?;
+                    let mut peekable = vals.items.iter().peekable();
+                    while let Some((name, type_)) = peekable.next() {
+                        write!(f, "{}: {}", name, type_)?;
+                        if peekable.peek().is_some() {
+                            write!(f, ", ")?;
+                        }
+                    }
+                    write!(f, ")")?;
+                }
+            }
+        };
+        Ok(())
+    }
+}
+
+impl Default for Results {
+    fn default() -> Self {
+        Results::empty()
+    }
+}
+
+impl From<Type> for Results {
+    fn from(value: Type) -> Self {
+        Results::Anon(value)
+    }
+}
+
+impl<N> FromIterator<(N, Type)> for Results
+where
+    N: Into<Ident>,
+{
+    fn from_iter<T: IntoIterator<Item = (N, Type)>>(iter: T) -> Self {
+        Results::Named(Params::from_iter(iter))
+    }
+}
+
+impl Results {
+    // For the common case of an empty results list.
+    pub fn empty() -> Results {
+        Results::Named(Default::default())
+    }
+
+    pub fn anon(type_: Type) -> Results {
+        Results::Anon(type_)
+    }
+
+    pub fn named(types: impl IntoIterator<Item = (impl Into<String>, Type)>) -> Results {
+        Results::Named(
+            types
+                .into_iter()
+                .map(|(name, ty)| (name.into(), ty))
+                .collect(),
+        )
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Results::Named(params) => params.items().len(),
+            Results::Anon(_) => 1,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+pub struct StandaloneFunction {
```

Comment:
Can we rename this to `Function`? That matches the term used in [wit-encoder](https://docs.rs/wit-parser/latest/wit_parser/struct.Function.html) too.

Reply:
We have StandaloneFunction  and ResourceFunc, whereas wit-parser has a single Function type that takes a kind argument.
I made separate types so that the type system enforces where each function kind can be. I.e. that standalone functions are not part of a resource, and method/static/constructor are not without a resource.

(also, should probably rename StandaloneFunction to StandaloneFunc)

Reply:
I  rename StandaloneFunction to StandaloneFunc